File: Leetcode/0071-simplify-path/solution.java
import java.util.*;

/**
 * 0071. Simplify Path
 * https://leetcode.com/problems/simplify-path/
 * Difficulty: Medium
 * Tags: String, Stack
 */
class Solution {

    /**
     * Optimal Solution (Split + Stack).
     * Time:  O(n)
     * Space: O(n)
     */
    public String simplifyPath(String path) {
        String[] parts = path.split("/");
        Deque<String> stack = new ArrayDeque<>();
        for (String p : parts) {
            if (p.isEmpty() || p.equals(".")) continue;
            if (p.equals("..")) {
                if (!stack.isEmpty()) stack.pop();
                continue;
            }
            stack.push(p);
        }
        StringBuilder sb = new StringBuilder();
        Iterator<String> it = stack.descendingIterator();
        while (it.hasNext()) sb.append('/').append(it.next());
        return sb.length() == 0 ? "/" : sb.toString();
    }

    public String simplifyPathManual(String path) {
        Deque<String> stack = new ArrayDeque<>();
        int i = 0, n = path.length();
        while (i < n) {
            while (i < n && path.charAt(i) == '/') i++;
            int j = i;
            while (j < n && path.charAt(j) != '/') j++;
            String seg = path.substring(i, j);
            i = j;
            if (seg.isEmpty() || seg.equals(".")) continue;
            if (seg.equals("..")) {
                if (!stack.isEmpty()) stack.pop();
                continue;
            }
            stack.push(seg);
        }
        StringBuilder sb = new StringBuilder();
        Iterator<String> it = stack.descendingIterator();
        while (it.hasNext()) sb.append('/').append(it.next());
        return sb.length() == 0 ? "/" : sb.toString();
    }

    static int passed = 0, failed = 0;
    static void test(String name, String got, String expected) {
        if (got.equals(expected)) { System.out.println("PASS: " + name); passed++; }
        else { System.out.println("FAIL: " + name + " got=" + got); failed++; }
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        String[][] cases = {
            {"Trailing slash", "/home/", "/home"},
            {"Up from root", "/../", "/"},
            {"Double slash", "/home//foo/", "/home/foo"},
            {"Mixed", "/a/./b/../../c/", "/c"},
            {"Just root", "/", "/"},
            {"Many ups", "/a/b/c/../../../", "/"},
            {"Hidden file", "/.hidden/file", "/.hidden/file"},
            {"Three dots", "/.../", "/..."},
            {"Multiple slashes", "//", "/"},
            {"Long names", "/abc/def/", "/abc/def"},
            {"Mixed with up at start", "/../abc/", "/abc"},
            {"Trailing trailing", "/abc//", "/abc"}
        };
        System.out.println("=== Split + Stack ===");
        for (String[] c : cases) test(c[0], sol.simplifyPath(c[1]), c[2]);
        System.out.println("\n=== Manual Parser ===");
        for (String[] c : cases) test("Manual " + c[0], sol.simplifyPathManual(c[1]), c[2]);
        System.out.printf("%nResults: %d passed, %d failed, %d total%n", passed, failed, passed + failed);
    }
}

File: Leetcode/0071-simplify-path/solution.py
# ============================================================
# 0071. Simplify Path
# https://leetcode.com/problems/simplify-path/
# Difficulty: Medium
# Tags: String, Stack
# ============================================================


class Solution:
    def simplifyPath(self, path: str) -> str:
        """
        Optimal Solution (Split + Stack).
        Time:  O(n)
        Space: O(n)
        """
        stack = []
        for p in path.split('/'):
            if p == '' or p == '.':
                continue
            if p == '..':
                if stack: stack.pop()
                continue
            stack.append(p)
        return '/' + '/'.join(stack)

    def simplifyPathManual(self, path: str) -> str:
        stack = []
        i, n = 0, len(path)
        while i < n:
            while i < n and path[i] == '/': i += 1
            j = i
            while j < n and path[j] != '/': j += 1
            seg = path[i:j]
            i = j
            if seg == '' or seg == '.': continue
            if seg == '..':
                if stack: stack.pop()
                continue
            stack.append(seg)
        return '/' + '/'.join(stack)


if __name__ == "__main__":
    sol = Solution()
    passed = failed = 0
    def test(name, got, expected):
        global passed, failed
        if got == expected: print(f"PASS: {name}"); passed += 1
        else: print(f"FAIL: {name} got={got!r}"); failed += 1

    cases = [
        ("Trailing slash", "/home/", "/home"),
        ("Up from root", "/../", "/"),
        ("Double slash", "/home//foo/", "/home/foo"),
        ("Mixed", "/a/./b/../../c/", "/c"),
        ("Just root", "/", "/"),
        ("Many ups", "/a/b/c/../../../", "/"),
        ("Hidden file", "/.hidden/file", "/.hidden/file"),
        ("Three dots", "/.../", "/..."),
        ("Multiple slashes", "//", "/"),
        ("Long names", "/abc/def/", "/abc/def"),
        ("Mixed with up at start", "/../abc/", "/abc"),
        ("Trailing trailing", "/abc//", "/abc"),
    ]

    print("=== Split + Stack ===")
    for n, p, exp in cases: test(n, sol.simplifyPath(p), exp)
    print("\n=== Manual Parser ===")
    for n, p, exp in cases: test("Manual " + n, sol.simplifyPathManual(p), exp)

    print(f"\nResults: {passed} passed, {failed} failed, {passed + failed} total")
